File: 05-Algorithms/code/python/seb_alg/bipartite_matching.py
"""Mirrors `typescript/src/bipartiteMatching.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence

from .errors import AlgoError


def bipartite_matching(n_left: int, n_right: int, edges: Sequence[Sequence[int]]) -> int:
    """Maximum bipartite matching size via Kuhn's algorithm (DFS-based
    augmenting paths, left vertices and each DFS's candidates both explored
    in ascending id order for determinism). O(V * E).
    """
    adj: List[List[int]] = [[] for _ in range(n_left)]
    for l, r in edges:
        if l < 0 or l >= n_left or r < 0 or r >= n_right:
            raise AlgoError("index", "left/right vertex out of bounds")
        adj[l].append(r)
    for lst in adj:
        lst.sort()

    match_right = [-1] * n_right

    def try_augment(l: int, visited: List[bool]) -> bool:
        for r in adj[l]:
            if visited[r]:
                continue
            visited[r] = True
            if match_right[r] == -1 or try_augment(match_right[r], visited):
                match_right[r] = l
                return True
        return False

    matched = 0
    for l in range(n_left):
        visited = [False] * n_right
        if try_augment(l, visited):
            matched += 1
    return matched

File: 05-Algorithms/code/typescript/src/bipartiteMatching.ts
import { AlgoError } from "./errors.js";

/**
 * Maximum bipartite matching size via Kuhn's algorithm (DFS-based
 * augmenting paths, one per left vertex, processed in ascending id order;
 * each DFS also tries right-vertex candidates in ascending id order, so the
 * specific matching found -- and therefore its size, which is all that's
 * asserted in vectors -- is deterministic). `edges` are `[left, right]`
 * pairs with `left` in `[0, nLeft)` and `right` in `[0, nRight)`. O(V * E).
 */
export function bipartiteMatching(nLeft: number, nRight: number, edges: [number, number][]): number {
  const adj: number[][] = Array.from({ length: nLeft }, () => []);
  for (const [l, r] of edges) {
    if (l < 0 || l >= nLeft || r < 0 || r >= nRight) throw new AlgoError("index", "left/right vertex out of bounds");
    adj[l].push(r);
  }
  for (const list of adj) list.sort((a, b) => a - b);

  const matchRight = new Array(nRight).fill(-1);

  function tryAugment(l: number, visited: boolean[]): boolean {
    for (const r of adj[l]) {
      if (visited[r]) continue;
      visited[r] = true;
      if (matchRight[r] === -1 || tryAugment(matchRight[r], visited)) {
        matchRight[r] = l;
        return true;
      }
    }
    return false;
  }

  let matched = 0;
  for (let l = 0; l < nLeft; l++) {
    const visited = new Array(nRight).fill(false);
    if (tryAugment(l, visited)) matched++;
  }
  return matched;
}
